libnetwork/drivers/overlay/bpf_test.go

```diff
@@ -12,3 +24,218 @@ func FuzzVNIMatchBPFDoesNotPanic(f *testing.F) {
 		_ = vniMatchBPF(vni)
 	})
 }
+
+func TestVNIMatchBPF(t *testing.T) {
+	// The BPF filter program under test uses Linux extensions which are not
+	// emulated by any user-space BPF interpreters. It is also classic BPF,
+	// which cannot be tested in-kernel using the bpf(BPF_PROG_RUN) syscall.
+	// The best we can do without actually programming it into an iptables
+	// rule and end-to-end testing it is to attach it as a socket filter to
+	// a raw socket and test which loopback packets make it through.
+	//
+	// Modern kernels transpile cBPF programs into eBPF for execution, so a
+	// possible future direction would be to extract the transpiler and
+	// convert the program under test to eBPF so it could be loaded and run
+	// using the bpf(2) syscall.
+	// https://elixir.bootlin.com/linux/v6.2/source/net/core/filter.c#L559
+	// Though the effort would be better spent on adding nftables support to
+	// libnetwork so this whole BPF program could be replaced with a native
+	// nftables '@th' match expression.
+	//
+	// The filter could be manually e2e-tested for both IPv4 and IPv6 by
+	// programming ip[6]tables rules which log matching packets and sending
+	// test packets loopback using netcat. All the necessary information
+	// (bytecode and acceptable test vector) is logged by this test.
+	//
+	//     $ sudo ip6tables -A INPUT -p udp -s ::1 -d ::1 -m bpf \
+	//         --bytecode '7,32 0 0 4294963252,7 0 0 0,64 0 0 4,116 0 0 8,21 1 0 4096,6 0 0 0,6 0 0 4294967295' \
```

Comment:
I'm not sure including the literal bytecode in the comment is a good idea, as future changes could cause it to become out of date. We could add a caveat here, or just substitute `[bytecode]` and some advice on how to use the Go assembler to output the bytecode for the above program.

Reply:
Fair enough. I assumed a savvy developer who could recognize the implication that this was merely a worked example since the VNI is constant is baked into the program, but that might not be a safe assumption.

Reply:
I don't think that's necessary at all if it shows up in the verbose log.